### DISCUSSION TITLE
Visualization of model evaluation (image, true mask, predicted mask)

### DISCUSSION THREAD

**Discussion by HripsimeS** [Project-MONAI/tutorials#607] [Q&A] [answered]
Hello. I am trying to do a visualization after the evaluation on original images/masks. For example I have 10 images/masks that I would like the test my model and then would like to plot results of all of them with images, true masks and predicted masks. The solution below does not work efficiently as it is plotting only one case (image, true mask and predicted mask) multiple times. Let me know if you have more efficient solution for this visualization. Look forward to hearing from you. 

loader = LoadImage()
original_image = loader(test_data[0]["image_meta_dict"]["filename_or_obj"])[0]

with torch.no_grad():
    for i, test_data in enumerate(test_loader):
        
        plt.figure("check", (18, 6))
        
        plt.subplot(1, 3, 1)
        plt.title(f"image {i}")
        plt.imshow(original_image[:, :, 250], cmap="gray")
        
        plt.subplot(1, 3, 2)
        plt.title(f"mask {i}")
        plt.imshow(test_labels[0].detach().cpu()[1, :, :, 250])
        
        plt.subplot(1, 3, 3)
        plt.title(f"prediction {i}")
        plt.imshow(test_outputs[0].detach().cpu()[1, :, :, 250])
        
        plt.show()
        if i == 2:
            break

P.S. If for evaluation of model I am using 10 images/masks, the metric_org is returning the average of all 10 scores? If yes, then how I can check dice score of each case individually?  This is the target part:

dice_metric(y_pred=test_outputs, y=test_labels)
metric_org = dice_metric.aggregate().item()
dice_metric.reset()

**Comment by yiheng-wang-nv** [comment, marked as answer]
Hi @HripsimeS , please put `original_image = loader(test_data[0]["image_meta_dict"]["filename_or_obj"])[0]` within your for loop:
```
for i, test_data in enumerate(test_loader):
    original_image = loader(test_data[0]["image_meta_dict"]["filename_or_obj"])[0]
   ...
```

**Reply by yiheng-wang-nv** [reply]
For the metric part, you can use `print(dice_metric(y_pred=val_outputs, y=val_labels))` if you'd like to see each individual score. Or you can use `dice_metric.get_buffer()` to check all calculated scores (to use this function, please remove the line `dice_metric.reset()`)

**Reply by HripsimeS** [reply]
@yiheng-wang-nv thanks a lot, dice metric part worked perfectly (both options). 
When I write original_image inside the for loop for visualization, I received error with this "KeyError: 0"

**Reply by yiheng-wang-nv** [reply]
Sorry I did not clarify it clearly. Please refer to https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/spleen_segmentation_3d.ipynb
If you do post processing for `test_data` first (shown as follow), it would be a list, thus it's necessary to achieve the dict `test_data[0]`.
```
test_data = [post_transforms(i) for i in decollate_batch(test_data)]
```
Otherwise, for a dict based data, you just need to do:
```
original_image = loader(test_data["image_meta_dict"]["filename_or_obj"])[0]
```

**Reply by HripsimeS** [reply]
@yiheng-wang-nv Thanks for your advice! It helped with images and it plots each unique image in the test list. But for test_labels and test_outputs I receive the plot of same case. When I remove [0] from test_labels and test_outputs, this error is appearing ('list' object has no attribute 'detach'). Removing "detach", it shows this error "list indices must be integers or slices, not tuple".

**Reply by yiheng-wang-nv** [reply]
can you post the full code here? I may need to check the data structure.

**Reply by HripsimeS** [reply]
@yiheng-wang-nv please find my notebook below, it is in the last part of notebook:
https://github.com/HripsimeS/Computer-Vision/blob/main/MONAI-Github.ipynb

Seems the loop is plotting true_mask and predicted_mask of only last case/image. And using dice_metric.get_buffer() I could check all calculated scores, but I had an impression for the finial dice_metric score it takes into account only the calculated scores for the last case. But if we upload 5 images let's say, I think would be good to get a mean/average of all results,

**Reply by yiheng-wang-nv** [reply]
you should insert your code into your 31th cell.

**Reply by HripsimeS** [reply]
@yiheng-wang-nv thank you! And what is wrong with data structure, why test_labels[0] and test_outputs[0] are not working efficiently inside of loop in "51st cell", eventually plotting results of the last case?

**Reply by yiheng-wang-nv** [reply]
In the 51st cell, test_labels and test_outputs are not defined. What you did is just to load the data:
```
with torch.no_grad():
    for i, test_data in enumerate(test_loader):
        loader = LoadImage()
        original_image = loader(test_data["image_meta_dict"]["filename_or_obj"])[0]
```

**Reply by HripsimeS** [reply]
@yiheng-wang-nv ok thanks. So I need to define them inside of the loop with the following code?

test_data = [post_transforms(i) for i in decollate_batch(test_data)]
test_outputs, test_labels = from_engine(["pred", "mask"])(test_data)

If I need to use this code inside of loop, then I received following error.
KeyError: 'Key `pred` of transform `EnsureTyped` was missing in the data and allow_missing_keys==False.'

Please let me know how would be better to define test_labels and test_outputs inside of loop to be able to see unique true and predicted masks of each case separately. Thanks a lot in advance for your help!

**Reply by yiheng-wang-nv** [reply]
HI @HripsimeS , I think the tutorial https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/spleen_segmentation_3d.ipynb is clear. Please check the last few cells carefully.

**Reply by HripsimeS** [reply]
@yiheng-wang-nv thank you very much for useful tips! I could succeed to plot all unique test cases with their image, true mask and predicted mask. I really appreciate a professional Q&A support of MONAI and if there is a place where I can leave my feedback, I will do with a great pleasure :) For other users, if you want to see a solution to do a visualization of all test cases you can find below: 

model.load_state_dict(torch.load(os.path.join(root_dir, "best_metric_model.pth")))
model.eval()
with torch.no_grad():
    for i, test_data in enumerate(test_loader):
        test_inputs = test_data["image"].to(device)
        roi_size = (160, 160, 160)
        sw_batch_size = 4
        test_data["pred"] = sliding_window_inference(test_inputs, roi_size, sw_batch_size, model)
        test_data = [post_transforms(i) for i in decollate_batch(test_data)]
        test_outputs, test_labels = from_engine(["pred", "mask"])(test_data)
        original_image = loader(test_data[0]["image_meta_dict"]["filename_or_obj"])[0]
        plt.figure("check", (18, 6))
        plt.subplot(1, 3, 1)
        plt.title(f"image {i}")
        plt.imshow(original_image[:, :, 250], cmap="gray")
        plt.subplot(1, 3, 2)
        plt.title(f"mask {i}")
        plt.imshow(test_labels[0].detach().cpu()[2, :, :, 250])
        plt.subplot(1, 3, 3)
        plt.title(f"prediction {i}")
        plt.imshow(test_outputs[0].detach().cpu()[2, :, :, 250])
        plt.show()
        if i == 4:
            break
